homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,142 @@
+        finally:
+            self.adb_lock = False
 
-    @property
-    def current_app(self):
-        """Return the current app."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'current'
-                    ), timeout=10).json()
-            _current_app = response.get('current_app')
-            if _current_app:
-                return _current_app.get('package')
-
-            return None
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve current app for %s", self.device_id)
-            return None
+        return returns
 
-    @property
-    def running_apps(self):
-        """Return a list of running apps."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'running'
-                    ), timeout=10).json()
-            return response.get('running_apps')
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve running apps for %s", self.device_id)
-            return None
-
-    def action(self, action_id):
-        """Perform an action on the device."""
-        try:
-            requests.get(DEVICE_ACTION_URL.format(
-                self.proto, self.host, self.port, self.device_id, action_id
-                ), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Action request for %s was not accepted for device %s",
-                action_id, self.device_id)
-
-    def start_app(self, app_name):
-        """Start an app."""
-        try:
-            requests.get(DEVICE_APPS_URL.format(
-                self.proto, self.host, self.port, self.device_id,
-                app_name + '/start'), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not start %s on %s", app_name, self.device_id)
+    return _adb_wrapper
 
 
 class FireTVDevice(MediaPlayerDevice):
     """Representation of an Amazon Fire TV device on the network."""
 
-    def __init__(self, proto, host, port, device, name):
+    def __init__(self, host, name, adbkey, get_source, get_sources):
         """Initialize the FireTV device."""
-        self._firetv = FireTV(proto, host, port, device)
+        from firetv import FireTV  # pylint: disable=no-name-in-module
+        from adb.adb_protocol import (
+            InvalidCommandError, InvalidResponseError, InvalidChecksumError)
+
+        self._host = host
         self._name = name
-        self._state = STATE_UNKNOWN
-        self._running_apps = None
+        self._adbkey = adbkey
+        self._get_source = get_source
+        self._get_sources = get_sources
+
+        self.firetv = FireTV(host, adbkey)
```

Comment:
Don't pass the adbkey if the user hasn't configured one. I would move the instantiation of the `FireTV` directly to `setup_platform` and check connection before instantiating the entity, if we don't want to continue with set up, if connection fails. Do we think that connection can succeed later?

Reply:
"Do we think that connection can succeed later?"

In general, people plug their Fire TV into an outlet (not into their TV), so it's never truly off. So if it fails to connect initially, then IMO it's safe to assume that there is an error in the configuration and the device should not be setup. If someone wants to later implement `PlatformNotReady`/retry logic later, that's fine. But as it is the current component is effectively broken, and until this working component gets integrated no additional work can be done.

Reply:
Fixed in https://github.com/home-assistant/home-assistant/pull/17767/commits/80daea67be3aa84a3e511aff54b95d7b3e4e18e7